ENH Add sparse input support to OPTICS

Micky774 opened the pull request:
#### Reference Issues/PRs
Fixes #11982
Resolves #14736 (stalled)
Resolves #20802 (stalled)

#### What does this implement/fix? Explain your changes.
PRs #14736 and #20802: Add support for sparse matrices in `OPTICS.fit`
This PR: updates work and provides continuation for review.

#### Any other comments?

huntzhan committed:
Try csr support.

huntzhan committed:
Change the default metric of OPTICS to euclidean.

huntzhan committed:
Retain the default metric minkowski.

huntzhan committed:
Undo tests.

huntzhan committed:
Fix flake8.

huntzhan committed:
Add sparse tests.

elefantinerrrrs committed:
Merged huntzhan

elefantinerrrrs committed:
Change assert_raise_message to pytest.raises

elefantinerrrrs committed:
Parametrized tests

elefantinerrrrs committed:
Fix flake8 test_optics.py

elefantinerrrrs committed:
Fix flake8 _optics.py

elefantinerrrrs committed:
Add sparse precomputed test case

elefantinerrrrs committed:
Black test_optics.py

elefantinerrrrs committed:
Merge remote-tracking branch 'upstream/main' into add-optics-sparse

elefantinerrrrs committed:
Black _optics.py

elefantinerrrrs committed:
Add sparse matrix support to _optics.py

Original commit was pushed, but wasn't reflected on github for some reason

elefantinerrrrs committed:
Merge remote-tracking branch 'origin/add-optics-sparse' into add-optics-sparse

elefantinerrrrs committed:
Added changelog entry

elefantinerrrrs committed:
Merge branch 'main' into add-optics-sparse

Micky774 committed:
Merge branch 'main' into sparse_optics

Micky774 committed:
Removed extra changelog entry

Micky774 committed:
Updated changelog entry

jeremiedbb commented on doc/whats_new/v1.0.rst:
```diff
@@ -499,6 +499,12 @@ Changelog
 - |API| :func:`cluster.spectral_clustering` raises an improved error when passed
   a `np.matrix`. :pr:`20560` by `Thomas Fan`_.
 
+- |Enhancement| The `predict` and `fit_predict` methods of
+  :class:`cluster.OPTICS` now accept sparse data type for input
+  data. :pr:`14736` by :user:`Hunt Zhan <huntzhan>`,
+  :pr:`20802` by :user:`Brandon Pokorny <Clickedbigfoot>`,
+  and :pr:`22965` by :user:`Meekail Zain <micky774>`
+
```
This should move to v1.1

jeremiedbb commented on sklearn/cluster/_optics.py:
```diff
@@ -285,7 +287,12 @@ def fit(self, X, y=None):
+        X = self._validate_data(X, dtype=dtype, accept_sparse="csr")
+        if self.metric == "precomputed" and issparse(X):
+            # Set each diagonal to an explicit value so each point is its own neighbor
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore", category=SparseEfficiencyWarning)
+                X.setdiag(X.diagonal())
```
Is it necessary ? Can't we let NearestNeighbors do whatever it needs to handle sparse precomputed matrix ?

In addition I think that the parameter description of ``metric`` should be updated to reflect the fact that it supports precomputed sparse graph, as done in the docstrings of nearest neighbors algorithms.

jeremiedbb commented on sklearn/cluster/tests/test_optics.py:
```diff
@@ -99,15 +101,26 @@ def test_extract_xi():
     X = np.vstack((C1, C2, C3, C4, C5, np.array([[100, 100]]), C6))
     expected_labels = np.r_[[2] * 5, [0] * 5, [1] * 5, [3] * 5, [1] * 5, -1, [4] * 5]
     X, expected_labels = shuffle(X, expected_labels, random_state=rng)
+    X = sparse.lil_matrix(X) if metric == "euclidean" else X
```
I don't understand that. Is it to test on dense and sparse ? or sparse doesn't work with minkowski ?

Micky774 replied:
Yes this was done due to sparse matrices not being supported by scipy distance metrics. I've included a clarifying line in the docstring description of `metric` to highlight this.

jeremiedbb commented on sklearn/cluster/tests/test_optics.py:
```diff
@@ -786,9 +876,11 @@ def test_extract_dbscan():
+def test_precomputed_dists(is_sparse):
     redX = X[::2]
     dists = pairwise_distances(redX, metric="euclidean")
+    dists = sparse.lil_matrix(dists).tocsr() if is_sparse else dists
```
```suggestion
    dists = sparse.csr_matrix(dists) if is_sparse else dists
```

jeremiedbb reviewed (commented):
Not sure what to think about these warnings. Maybe we should not filter them since they indicate that the use of NearestNeighbors is not optimal. At least maybe there's something we can do to the precomputed dist matrix to not trigger all of them.

I think not all tests need to be parametrized by metric and run on dense and sparse, especially those who just do error checking or sanity checks.

Micky774 committed:
Update changelog

Micky774 committed:
Streamlined tests

Micky774 committed:
Added disclaimer on sparse matrix support for metrics

Micky774 committed:
Removed sparse parametrization for algorithm edge-case test

Micky774 commented:
@jeremiedbb I have
1. Converted `lil_matrix` constructions to `csr_matrix`
2. Improved the docstring for `metric` by including a line stating that `Sparse matrices are only supported by scikit-learn metrics.`
3. Removed the warning filters so that they are passed regularly
4. Streamlined the tests by removing sparse parametrization for several tests, keeping the parametrization only for testing the correctness of the algorithm, DBSCAN parity, and the ability to input sparse precomputed distances.

Micky774 committed:
Merge branch 'main' into sparse_optics

Micky774 committed:
Remove unused metric argument

Micky774 committed:
Cleanup to minimize diff

Micky774 committed:
Merge branch 'main' into sparse_optics

Micky774 committed:
Merge branch 'main' into sparse_optics

Micky774 commented:
Note that the failing test is `test_ridge_regression_vstacked_X[wide-5-True-sag]` and consequently not related to this PR

jeremiedbb reviewed (commented):
Looks good. I just think that you added some unreachable code, but that needs confirmation

Micky774 commented on sklearn/cluster/_optics.py:
```diff
@@ -609,12 +615,17 @@ def _set_reach_dist(
 
-    rdists = np.maximum(dists, core_distances_[point_index])
-    np.around(rdists, decimals=np.finfo(rdists.dtype).precision, out=rdists)
+    if issparse(dists):
```
Yup, just checked and it is guaranteed to be dense -- good catch, thanks

Micky774 replied:
Looks like it's actually to catch the case that `metric='precomputed'` and is passed a sparse matrix, which is a valid concern, so I don't think it needs to be changed.

jeremiedbb replied:
right, I missed that

Micky774 committed:
Merge branch 'main' into sparse_optics

jeremiedbb commented on sklearn/cluster/_optics.py:
```diff
@@ -609,12 +615,17 @@ def _set_reach_dist(
-    rdists = np.maximum(dists, core_distances_[point_index])
-    np.around(rdists, decimals=np.finfo(rdists.dtype).precision, out=rdists)
+    if issparse(dists):
+        rdists = dists.maximum(core_distances_[point_index])
```
Scipy warns here that max will result in a dense matrix. Since we convert to dense anyway it would make sense to convert before this. I'd convert at line 611:
```py
    if metric == "precomputed":
        dists = X[point_index, unproc]
        if issparse(dists):
            dists = dists.toarray().ravel()
```

Then, this branch is no longer useful

jeremiedbb commented on sklearn/cluster/_optics.py:
```diff
@@ -285,7 +288,10 @@ def fit(self, X, y=None):
-        X = self._validate_data(X, dtype=dtype)
+        X = self._validate_data(X, dtype=dtype, accept_sparse="csr")
+        if self.metric == "precomputed" and issparse(X):
+            # Set each diagonal to an explicit value so each point is its own neighbor
+            X.setdiag(X.diagonal())
```
Scipy complains that this is inefficient for CSR format. Should we validate with ``accept_sparse="lil" if self.metric == "precomputed" else "csr"`` and then convert to csr ?

Micky774 replied:
Interestingly enough, it seems as though `lil` (as well as `dok`) don't seem to be able to accept explicit zero-values which is what needs to happen to have the graph built correctly. We could manually do it, but that'll probably end up even worse than the current implementation for any decently sized matrix. 

I may just be missing an obvious way to do this since I'm unfamiliar with manipulating scipy sparse matrices, but afaik this is a necessary evil right now.

thomasjpfan replied:
I do not think any estimators with `metric="precomputed"` support precomputed sparse data. In practice, I do not think precomputed distances are usually sparse.

With that in mind, I would be happy with raising an error for `metric == "precomputed"` and sparse data.

jeremiedbb replied:
maybe we can filter this warning and add a small comment.

thomasjpfan replied:
I am okay with filtering, since we do the same thing in DBSCAN:

https://github.com/scikit-learn/scikit-learn/blob/f9a82d28e7f0f0b1f274acad9607eaa38fb12bc8/sklearn/cluster/_dbscan.py#L355-L360

Micky774 committed:
Update sklearn/cluster/_optics.py

Co-authored-by: Jérémie du Boisberranger <34657725+jeremiedbb@users.noreply.github.com>

Micky774 committed:
Merge branch 'main' into sparse_optics

Micky774 committed:
Added explicit dense cast before evaluating maximal distances.

Micky774 committed:
Merge branch 'main' into sparse_optics

jeremiedbb commented on doc/whats_new/v1.1.rst:
```diff
@@ -206,6 +206,12 @@ Changelog
   `-1` and the original warning message is shown.
   :pr:`22217` by :user:`Meekail Zain <micky774>`.
 
+- |Enhancement| The `predict` and `fit_predict` methods of
+  :class:`cluster.OPTICS` now accept sparse data type for input
+  data. :pr:`14736` by :user:`Hunt Zhan <huntzhan>`,
+  :pr:`20802` by :user:`Brandon Pokorny <Clickedbigfoot>`,
+  and :pr:`22965` by :user:`Meekail Zain <micky774>`.
+
```
it should move to 1.2 now

jeremiedbb reviewed (approved):
LGTM besides the sparse efficiency warning thing.

Micky774 committed:
Merge branch 'main' into sparse_optics

Micky774 committed:
Surpressed unnecessary warnings

thomasjpfan commented on sklearn/cluster/_optics.py:
```diff
@@ -603,15 +612,15 @@ def _set_reach_dist(
     if metric == "precomputed":
         dists = X[point_index, unproc]
+        if issparse(dists):
+            dists = dists.toarray().ravel()
```
I do not think we need to make this dense:

```suggestion
            dists.sort_indices()
            dists = dists.data
```

It is a trade off between sorting the indices and densifying the sparse matrix. In the sparse precomputed metric use case, I suspect there are not many indices and densifying is more expensive.

Micky774 committed:
Merge branch 'main' into sparse_optics

Micky774 committed:
Update sklearn/cluster/_optics.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

thomasjpfan commented on sklearn/cluster/_optics.py:
```diff
@@ -603,15 +612,16 @@ def _set_reach_dist(
     if metric == "precomputed":
         dists = X[point_index, unproc]
+        if issparse(dists):
+            dists.sort_indices()
+            dists = dists.data
```
@jeremiedbb This is a change from your previous approval. Are you okay with this?

thomasjpfan reviewed (approved):
LGTM

jeremiedbb committed:
fix position in changelog

jeremiedbb commented:
Thanks @Micky774 !
